TESTS/network/wifi/wifi_connect_nonblock.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2019, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed.h"
+#include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
+#include "wifi_tests.h"
+
+using namespace utest::v1;
+
+#if defined(MBED_CONF_APP_WIFI_UNSECURE_SSID)
+
+#define SSID_MAX_LEN 32
+
+void wifi_connect_nonblock(void)
+{
+    WiFiInterface *wifi = get_interface();
+
+    TEST_SKIP_UNLESS(wifi-> set_blocking(false) != NSAPI_ERROR_UNSUPPORTED)
+
+    char ssid[SSID_MAX_LEN + 1] = MBED_CONF_APP_WIFI_UNSECURE_SSID;
+
+    TEST_ASSERT_EQUAL_INT(NSAPI_ERROR_OK, wifi->set_credentials(ssid, NULL));
+
+    nsapi_error_t ret = wifi->connect();
+
+    TEST_ASSERT_TRUE((ret == NSAPI_ERROR_OK) || (ret == NSAPI_ERROR_WOULD_BLOCK));
+
+    nsapi_error_t connection_status;
+    if (ret == NSAPI_ERROR_OK) {
+        connection_status = wifi->get_connection_status();
+        TEST_ASSERT_TRUE(
+            (connection_status == NSAPI_STATUS_GLOBAL_UP)
+            || (connection_status == NSAPI_STATUS_LOCAL_UP))
+    } else {
+        while ((connection_status != NSAPI_STATUS_GLOBAL_UP)
+                && (connection_status != NSAPI_STATUS_LOCAL_UP)) {
+            connection_status =  wifi->get_connection_status();
```

Comment:
You are effectively polling in busy-loop.
Please add some kind of delay here, or use callbacks that report connection status.

Reply:
Asynchronous execution added, @Seppo, please review again.

Reply:
Asynchronous execution added, @SeppoTakalo , please review again.